2_process/src/lake_sequences_mntoha.py

```diff
@@ -0,0 +1,228 @@
+        # only the first will be used.
+        # If there are no matches, an IndexError will be raised.
+        lake = lake_metadata_augmented[lake_metadata_augmented.site_id==site_id].iloc[0,:]
+
+        # Create one large DataFrame of temperatures at all days and all discretized depths
+        obs_full = all_dates_depths(lake_obs_interpolated,
+                                    depths,
+                                    pad_before_days=spinup_time,
+                                    min_days=sequence_length)
+        # Now obs_full has dimensions of len(date_range) by len(depths)
+        # Values are NaN wherever/whenever there are no observations
+
+        # There are some NaN values in the clarity and ice flag data.
+        # There are multiple ways to deal with NaN values in inputs:
+        # interpolate to fill them, skip over them, or remove sequences
+        # containing NaNs entirely. At this point, allow NaN values into the
+        # sites. Later code will handle NaN values.
+
+        # Join drivers by date
+        drivers = pd.read_csv(drivers_file, parse_dates=['time'], index_col='time')
+        obs_full = obs_full.join(drivers, how='left')
+
+        # Join clarity by date
+        clarity = pd.read_csv(clarity_file, parse_dates=['date'], index_col='date')
+        obs_full = obs_full.join(clarity, how='left')
+
+        # Join ice flags by date
+        ice_flags = pd.read_csv(ice_flags_file, parse_dates=['date'], index_col='date')
+        obs_full = obs_full.join(ice_flags, how='left')
+
+        # Get attributes
+        obs_full['area'] = lake['area']
+        obs_full['lon'] = lake['centroid_lon']
+        obs_full['lat'] = lake['centroid_lat']
+        obs_full['elevation'] = lake['elevation']
+        # Now obs_full is one long timeseries with depth-specific observations,
+        # drivers, clarity, ice_flags, and attributes as columns
+
+        # Split into sequences
+
+        # Let's make some rules for sequences:
+        # 1. Sequences are sequence_length days long
+        # 2. Each next sequence starts sequence_offset days later than the previous
+        # 3. Discard sequences that lack observations
+        # 4. We know going in that the first and last sequences have observations
+
+        # Convert dataframe to numpy array
+        obs_full_array = obs_full.to_numpy(dtype=np.float32)
+        # Create a strided view into obs_full_array
+        # This creates a numpy object with sequences of length sequence_length
+        # and offsets of sequence_offset, without making an in-memory copy of the array
+        all_sequences = np.lib.stride_tricks.sliding_window_view(
+            obs_full_array, sequence_length, axis=0
+        )[::sequence_offset]
+        # The shape of all_sequences is (# sequences, # features + depths, sequence_length)
+        # The first `len(depths)` elements in the middle dimension are observations
```

Comment:
since `depths` come first in that second dimension, I'd change all references to that dimension, e.g., line 164, to make it clear that depth values precede the features in that dimension, rather than the other way around. Also I'd just clarify that you mean _temperature_ observations here

Reply:
Yes, thanks! I had the ordering of depths and features backwards in my head at one time, and never updated the comments. Observations clarified to mean temperatures, here and elsewhere.